### PR TITLE
Add NLATable analysis and use it in PrefixModules

### PR DESCRIPTION
**Pull request by darthscsi**

### CODE REVIEW

**Review by fabianschuiki — APPROVED** [review]
This looks fantastic. The Grand Central Data Taps pass will benefit a lot from having `NLATable`s around! Can't wait to make use of this.

File: lib/Dialect/FIRRTL/NLATable.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// void NLATable::replaceInstance(InstanceOp oldInst, InstanceOp newInst) {
+//     assert(oldInst.moduleName() == newInst.moduleName() && "module targets
+//     mismatch"); assert(oldInst->getParentOfType<FModuleOp>() ==
+//     newInst->getParentOfType<FModuleOp>() && "module targets mismatch"); auto
+//     nlas = nodeMap.equal_range(oldInst->getParentOfType<FModuleOp>()); for
+//     (auto i = nlas.first, e = nlas.second; i != e; ++i)
+//         i->second.updateInstance(oldInst, newInst);
+// }
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/NLATable.cpp:48]: Should this be part of the public API?

File: test/Dialect/FIRRTL/prefix-modules.mlir [hunk 2]
```diff
@@ -291,7 +291,7 @@ firrtl.circuit "GCTDataMemTapsPrefix" {
 // Test the the NonLocalAnchor is properly updated.
 // CHECK-LABEL: firrtl.circuit "FixNLA" {
   firrtl.circuit "FixNLA"   {
-    firrtl.nla @nla_1 [#hw.innerNameRef<@FixNLA::@bar>, #hw.innerNameRef<@Bar::@baz>, @Baz]
+     firrtl.nla @nla_1 [#hw.innerNameRef<@FixNLA::@bar>, #hw.innerNameRef<@Bar::@baz>, @Baz]
```
Comment (fabianschuiki) [inline test/Dialect/FIRRTL/prefix-modules.mlir:294]: Stray space

**Review by dtzWill — COMMENTED** [review]
Hi! Just some small things I noticed doing a fly-by and checking things out, hope that's alright :+1:. (sorry for noise otherwise!)

File: lib/Dialect/FIRRTL/NLATable.cpp (hunk 1 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/NLATable.cpp:48]: Yes, but I don't have a user of it yet, so deleted.  Didn't mean to leave it in :)

File: test/Dialect/FIRRTL/prefix-modules.mlir (hunk 2 above)
Comment (darthscsi) [reply test/Dialect/FIRRTL/prefix-modules.mlir:294]: where is mlir-format when you need it?

File: include/circt/Dialect/FIRRTL/NLATable.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  /// CircuitOp or MLIR ModuleOp.
+  explicit NLATable(Operation *operation);
+
+  /// Lookup all NLAs an operation participates in
+  ArrayRef<NonLocalAnchor> lookup(Operation *op);
+
+  /// Lookup all NLAs an operation participates in
+  ArrayRef<NonLocalAnchor> lookup(StringAttr name);
+
+  /// Resolve an symbol to an NLA
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/NLATable.h:40]: where is english-tidy when you need it?

File: lib/Dialect/FIRRTL/NLATable.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+void NLATable::renameModule(StringAttr oldModName, StringAttr newModName) {
+  auto op = symToOp.find(oldModName);
+  if (op == symToOp.end())
+    return;
+  auto iter = nodeMap.find(oldModName);
+  if (iter == nodeMap.end())
+    return;
+  for (auto nla : iter->second)
+    nla.updateModule(oldModName, newModName);
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply lib/Dialect/FIRRTL/NLATable.cpp:48]: It's not actually dropping references form the map, but from the vector in the map.  It should drop an entry from the map if the vector becomes empty though.

File: test/Dialect/FIRRTL/prefix-modules.mlir (hunk 2 above)
Comment (fabianschuiki) [reply test/Dialect/FIRRTL/prefix-modules.mlir]: :+1: for mlir-format!

**Pull request merged**